### PR TITLE
Allows generation of redirects.json to be disabled

### PR DESCRIPTION
**Pull request by dmalan**
Per https://github.com/jekyll/jekyll-redirect-from/issues/189, it's not always desirable to generate [`redirects.json`](https://github.com/jekyll/jekyll-redirect-from/pull/147), particularly since it might leak information about routes that exist but that aren't actually (intended to be) linked publicly.

This PR proposes to add support for

```yml
redirect_from:
  json: false
```

modeled after https://github.com/benbalter/jekyll-optional-front-matter#configuration.

### CODE REVIEW

File: lib/jekyll-redirect-from/generator.rb [hunk 1]
```diff
@@ -5,6 +5,9 @@ class Generator < Jekyll::Generator
     safe true
     attr_reader :site, :redirects
 
+    CONFIG_KEY = "redirect_from".freeze
+    JSON_KEY = "json".freeze
```
Comment (ashmaroli) [inline lib/jekyll-redirect-from/generator.rb]: Please remove these constants.
This plugin already requires at least Ruby 2.3 and has all string literals frozen by default.

Calling `"redirect_from"` even a 100 times is not going to impact performance as it'd return the same Ruby object each time.

File: lib/jekyll-redirect-from/generator.rb [hunk 2]
```diff
@@ -22,11 +25,17 @@ def generate(site)
         generate_redirect_to(doc)
       end
 
-      generate_redirects_json
+      if generate_redirects_json?
+        generate_redirects_json
```
Comment (ashmaroli) [inline lib/jekyll-redirect-from/generator.rb]: Please inline this with a shorter check:
```ruby
generate_redirects_json if generate_json?
```

File: lib/jekyll-redirect-from/generator.rb [hunk 3]
```diff
@@ -22,11 +25,17 @@ def generate(site)
...
-      generate_redirects_json
+      if generate_redirects_json?
+        generate_redirects_json 
+      end
     end
 
     private
 
+    def option(key)
+      site.config[CONFIG_KEY] && site.config[CONFIG_KEY][key]
```
Comment (ashmaroli) [inline lib/jekyll-redirect-from/generator.rb]: Prefer [`Hash#dig`](https://ruby-doc.org/core-2.3.0/Hash.html#method-i-dig) from Ruby 2.3 here:
```suggestion
      site.config.dig("redirect_from", key)
```

File: lib/jekyll-redirect-from/version.rb [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # frozen_string_literal: true
 
 module JekyllRedirectFrom
-  VERSION = "0.15.0"
+  VERSION = "0.16.0"
```
Comment (ashmaroli) [inline lib/jekyll-redirect-from/version.rb]: Please **undo this change**.
The version will be set at the time of release.

File: spec/jekyll_redirect_from/generator_spec.rb [hunk 5]
```diff
@@ -94,55 +94,67 @@
...
 
-    it "creates the redirets file" do
-      expect(path).to exist
-    end
-
-    it "contains redirects" do
-      expect(redirects.count).to eql(13)
+    context "when explicitly disabled" do
+      let(:site) do
+        fixture_site("site", { "redirect_from" => { "json" => false } })
```
Comment (ashmaroli) [inline spec/jekyll_redirect_from/generator_spec.rb]: Where is this `fixture_site` method defined?

Comment (dmalan) [reply spec/jekyll_redirect_from/generator_spec.rb]: This is at https://github.com/jekyll/jekyll-redirect-from/pull/207/files#diff-93830fa29d616f7c87903d08b5b1b29aR65, modeled after https://github.com/benbalter/jekyll-optional-front-matter/blob/master/spec/spec_helper.rb#L26-L35.

**Comment by dmalan** [comment]
Changes made!

File: spec/jekyll_redirect_from/generator_spec.rb (hunk 5 above)
Comment (ashmaroli) [reply spec/jekyll_redirect_from/generator_spec.rb]: Whoa! I did not see that file before.. :stuck_out_tongue: 
Additionally, there already seems to be a `fixture_path` [helper method](https://github.com/jekyll/jekyll-redirect-from/pull/207/files#diff-93830fa29d616f7c87903d08b5b1b29aL24).
Can that be reused..?

Comment (dmalan) [reply spec/jekyll_redirect_from/generator_spec.rb]: Looks like it's only used by `fixture_site` (in [jekyll-optional-front-matter](https://github.com/benbalter/jekyll-optional-front-matter/blob/master/spec/spec_helper.rb#L26-L35) too), so happy to integrate its one-liner into `fixture_site` itself? Didn't want to drift from any existing paradigms in GitHub Pages-official plugins is all!

Comment (ashmaroli) [reply spec/jekyll_redirect_from/generator_spec.rb]: The set ups are different. For example, there is no fixture directory named `"site"` in this repo or in this PR. So your spec is fundamentally flawed.

File: lib/jekyll-redirect-from/generator.rb [hunk 6]
```diff
@@ -22,11 +22,15 @@ def generate(site)
         generate_redirect_to(doc)
       end
 
-      generate_redirects_json
+      generate_redirects_json if generate_redirects_json?
     end
 
     private
 
+    def option(key)
```
Comment (ashmaroli) [inline lib/jekyll-redirect-from/generator.rb]: Lets do away with this method for now. We can introduce it in the future if necessary.

File: lib/jekyll-redirect-from/generator.rb [hunk 7]
```diff
@@ -53,5 +57,9 @@ def generate_redirects_json
       page.data["layout"] = nil
       site.pages << page
     end
+
+    def generate_redirects_json?
+      option("json") != false
```
Comment (ashmaroli) [inline lib/jekyll-redirect-from/generator.rb]:
```suggestion
      site.config.dig("redirect_from", "json") != false
```

File: spec/jekyll_redirect_from/generator_spec.rb (hunk 5 above)
Comment (dmalan) [reply spec/jekyll_redirect_from/generator_spec.rb]: Ah, quite fair. How is https://github.com/jekyll/jekyll-redirect-from/pull/207/commits/95d21757094313ee9f6f634bfb6e0e4b2dccf5a2 as an alternative? Afraid this is my first time with ruby/spec, but as best I can tell, this would seem to override `site.config`?

Comment (ashmaroli) [reply spec/jekyll_redirect_from/generator_spec.rb]: Yes, that would override the `site.config`, but then it wouldn't quite be a proper configuration because the site's source points elsewhere. Here's what you need to do:
- In `spec/spec_helper.rb`, modify `site` to take an optional override:
```ruby
  def site(overrides = {})
    @site ||= Jekyll::Site.new(config.merge(overrides))
  end
```
- Use it in your test:
```ruby
let(:custom_site) { site("json" => false) }
...
```
*Note: The above is untested. However, you get the idea..*

Comment (dmalan) [reply spec/jekyll_redirect_from/generator_spec.rb]: Hm, so with `site` redefined in that way, and a test like

```ruby
context "when explicitly disabled" do
  let(:custom_site) { site("redirect_from" => { "json" => false }) }
  it "does not create the redirects file" do
    puts custom_site.config.to_yaml
    ...
  end
end
```

it looks like `custom_site.config` ends up being identical to `site.config` (i.e., the overrides are missing), presumably because `site("redirect_from" => { "json" => false })` ends up returning `@site`, which is already set by some earlier invocation of `site()`, perhaps because of the outermost [`before`](https://github.com/jekyll/jekyll-redirect-from/blob/master/spec/jekyll_redirect_from/generator_spec.rb#L4-L9)? Am I perhaps misunderstanding what you had in mind?

Thanks!

Comment (ashmaroli) [reply spec/jekyll_redirect_from/generator_spec.rb]: My bad.. The reason you're seeing this is because `:site` is a memoized method. It returns the same `@site` once the instance variable is set to a non-false, non-nil value.

Sorry abt that.

So I guess, the alternative route would be to:
```ruby
let(:custom_site) { Jekyll::Site.new(config.merge("json" => false)) }
...
```

Comment (dmalan) [reply spec/jekyll_redirect_from/generator_spec.rb]: Ah, no worries. Would it be incorrect to override `site` itself, as with https://github.com/jekyll/jekyll-redirect-from/pull/207/commits/6c0f2791f1632b7f75982f977fede83c9b81a873? If a new symbol like `custom_site` is indeed necessary instead, is an additional `before` necessary (e.g., to call `custom_site.read`, `custom_site.generate`, `custom_site.render`, and/or `custom_site.write) or any other such setup?

Thanks!

**Comment by dmalan** [comment]
Thanks!

File: README.md [hunk 8]
```diff
@@ -127,6 +127,19 @@ Your layout will get the following variables:
...
+
+You can configure this plugin in `_config.yml` by adding to the `redirect_from` key.
+
+### Disabling `redirects.json`
+
+By default, a file called `redirects.json`, which can be used for automated testing or to implement server-side redirects, will be included in the output. To exclude it from the output, set the `json` key to `false`:
+
+```yml
+redirect_from:
+  json: false
```
Comment (DirtyF) [inline README.md]: `output: false` would have been better here.

Comment (parkr) [reply README.md]: `output` feels like every redirect file! I think we should be more specific. `json` works for me.

**Comment by DirtyF** [comment]
@jekyllbot: merge +minor
